scipy/spatial/distance.pyi

```diff
@@ -0,0 +1,226 @@
+import sys
+from typing import overload, Optional, Any, Union, Tuple, SupportsFloat
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+if sys.version_info >= (3, 8):
+    from typing import Literal, Protocol, SupportsIndex
+else:
+    from typing_extensions import Literal, Protocol
+
+# Anything that can be parsed by `np.float64.__init__` and is thus
+# compatible with `ndarray.__setitem__` (for a float64 array)
+if sys.version_info >= (3, 8):
+    _FloatValue = Union[None, str, bytes, SupportsFloat, SupportsIndex]
+else:
+    _FloatValue = Union[None, str, bytes, SupportsFloat]
+
+class _MetricCallback1(Protocol):
+    def __call__(
+        self, __XA: np.ndarray, __XB: np.ndarray
+    ) -> _FloatValue: ...
+
+class _MetricCallback2(Protocol):
+    def __call__(
+        self, __XA: np.ndarray, __XB: np.ndarray, **kwargs: Any
+    ) -> _FloatValue: ...
+
+# TODO: Use a single protocol with a parameter specification variable
+# once available (PEP 612)
+_MetricCallback = Union[_MetricCallback1, _MetricCallback2]
+
+_MetricKind = Literal[
+    'braycurtis',
+    'canberra',
+    'chebychev', 'chebyshev', 'cheby', 'cheb', 'ch',
+    'cityblock', 'cblock', 'cb', 'c',
+    'correlation', 'co',
+    'cosine', 'cos',
+    'dice',
+    'euclidean', 'euclid', 'eu', 'e',
+    'matching', 'hamming', 'hamm', 'ha', 'h',
+    'minkowski', 'mi', 'm', 'pnorm',
+    'jaccard', 'jacc', 'ja', 'j',
+    'jensenshannon', 'js',
+    'kulsinski',
+    'mahalanobis', 'mahal', 'mah',
+    'rogerstanimoto',
+    'russellrao',
+    'seuclidean', 'se', 's',
+    'sokalmichener',
+    'sokalsneath',
+    'sqeuclidean', 'sqe', 'sqeuclid',
+    # NOTE: deprecated
+    # 'wminkowski', 'wmi', 'wm', 'wpnorm',
+    'yule',
+]
+
+# Function annotations
+
+def braycurtis(
+    u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
+) -> np.float64: ...
+
+def canberra(
+    u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
+) -> np.float64: ...
+
+# TODO: Add `metric`-specific overloads
+@overload
+def cdist(
+    XA: ArrayLike,
+    XB: ArrayLike,
+    metric: _MetricKind = ...,
+    *,
+    out: Optional[np.ndarray] = ...,
+    p: float = ...,
+    w: Optional[ArrayLike] = ...,
+    V: Optional[ArrayLike] = ...,
+    VI: Optional[ArrayLike] = ...,
+) -> np.ndarray: ...
+@overload
+def cdist(
+    XA: ArrayLike,
+    XB: ArrayLike,
+    metric: _MetricCallback,
+    *,
+    out: Optional[np.ndarray] = ...,
+    **kwargs: Any,
+) -> np.ndarray: ...
+
+# TODO: Wait for dtype support; the return type is
+# dependent on the input arrays dtype
```

Comment:
`chebyshev` and `cityblock` are documented as `double`. Looking at the implementation, I guess they're also returning `int`? I'd be fine with just returning `float` (or perhaps better, `np.float64`) for these, rather than making the type annotations very complex.

Reply:
Yeah, if I were to speculate then I'd say ~90% of all use cases return either `np.int_` or `np.float64`.
We can either change the return-type to one these two or just stick with `Any` for now.

As a side note: if we're going the `np.float64` route then I'd propose changing the function's example section a bit.
Namely, to change the example int-based array into a float-based one.